### ISSUE TITLE
Automatically change localhost references in HostAndPort to the address of the current machine

### ISSUE THREAD

**Issue by allanwax** [xetorthio/jedis#1023]
First, apologies as I'm still sick (today=2015-06-11) so if there are stupid errors below, please forgive.

Many people get caught specifying localhost/127.0.0.1 in configuration values for jedis cluster.  It is fine for jedis.  I myself got caught when I first started using it.  I propose to help by automatically replacing localhost references to the IP address of the local machine.  Thus as information gets passed around it will contain an absolute address rather than a relative one.

Here's the suggested new class.  NOTE `169.254` is a Microsoft special address sometimes used if the machine can't get to the internet.  I.e. the connections window will show a connection to the router but none to the internet.

``` java
package redis.clients.jedis;

import java.net.InetAddress;

public class HostAndPort {
    public static final String LOCALHOST_STR ;

    static {
        try {
            LOCALHOST_STR = InetAddress.getLocalHost().getAddress();
        }
        catch (Exception e) {
            LOCALHOST_STR = "localhost";
        }
    }

    private String host;
    private int port;

    public HostAndPort(String host, int port) {
        this.host = host;
        this.port = port;
    }

    public String getHost() {
        return host;
    }

    public int getPort() {
        return port;
    }

    @Override
    public boolean equals(Object obj) {
                if (obj == null) return false;

        if (obj instanceof HostAndPort) {
            HostAndPort hp = (HostAndPort) obj;
            String thisHost = convertHost(host);
            String hpHost = convertHost(hp.host);
            return port == hp.port && thisHost.equals(hpHost);
        }

        return false;
    }

    @Override
    public int hashCode() {
        return 31 * convertHost(host).hashCode() + port;
    }

    @Override
    public String toString() {
        return host + ":" + port;
    }

    private String convertHost(String host) {
         if (host.equals("127.0.0.1") || host.startsWith("localhost") || host.equals("0.0.0.0") || host.startsWith("169.254"))
             return LOCALHOST_STR;
         else if (host.startsWith("::1") || host.startsWith("0:0:0:0:0:0:0:1"))
            return LOCALHOST_STR;

        return host;
    }
}
```

**Comment by marcosnils** [comment]
@allanwax you already did all the hard work. Why not submitting a PR?.
I agree it's something it will definitely help Jedis users.

**Comment by allanwax** [comment]
I'm not a GIT person, so I'd probably screw up the repo trying.  Plus my Eclipse instance is giving me problems.  I'd appreciate it if you'd check it in.

**Comment by marcosnils** [comment]
@allanwax sure. I'll create the PR and give you credit for it :smile:

**Comment by allanwax** [comment]
Thanks!

**Comment by marcosnils** [comment]
@allanwax I've been looking at this change and I don't quite understand how we'll solve the absolute variable definition as the `convert` method in the class you're proposing to change is only called from the `equal` and `hashcode` methods in the same class. 

Looking a bit deeper in the code seems like no-one uses the `convert` method.

**Comment by allanwax** [comment]
Missed it.  Try this change

```java
public HostAndPort(String host, int port) {
    this.host = convertHost(host);
    this.port = port;
  }
```

then every time a HostAndPort is created it will be processed before being stored.

**Comment by allanwax** [comment]
If `setHost` is implemented, it will have to do the same thing.

**Comment by allanwax** [comment]
With the above change, `String thisHost = convertHost(host); ... return port == hp.port && thisHost.equals(hpHost);` in equals can just be `return port == hp.port && host.equals(hpHost);`

**Comment by allanwax** [comment]
@marconils I was wondering if this change was incorporated and tested

**Comment by marcosnils** [comment]
@allanwax haven't had the time to include it yet. Maybe you can become friend with git and submit a PR :wink:

**Comment by allanwax** [comment]
@marcosnils As my first experiment in git I have the above changes in my own branch.  I can't figure out how to push it up for you to examine to then be merged into jedis.   newbie assistance requested.

**Comment by marcosnils** [comment]
@allanwax this are the steps you need to do to submit a PR

1. Fork the jedis repo into your own github account using the fork github button
![image](https://cloud.githubusercontent.com/assets/1578458/9019712/7fe7a5f4-37c6-11e5-96bd-26685ad71565.png)

2. Add your repo as a remote of your current Jedis project with `git remote add allanwax git@github.com:allanwax/jedis`

3. Push your branch to __your__ repo with `git push allanwax <branchname>`

4. Go to your repo github page, switch to the branch you've pushed and click the `pull request` option: 

![image](https://cloud.githubusercontent.com/assets/1578458/9019725/d3da30a0-37c6-11e5-8192-fa3eb88dcea2.png)


That will open a pull request against this repo. Let me know if you have any issues

**Comment by sazzad16** [comment]
Closed by #1342 and #1468
